File: 2017/day03_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import math
from typing import Any


def solution(content: int) -> Any:
    if content < 1:
        raise ValueError(f"Invalid input: {content}")
    if content == 1:
        return 0
    if content < 10:
        return 1 if content % 2 == 0 else 2
    side = math.ceil(content**0.5)
    side = side if side % 2 == 1 else side + 1
    half = side // 2
    diff = min(
        abs(x - content)
        for x in range(side * side - half, (side - 2) * (side - 2), 1 - side)
    )
    return half + diff


def main() -> None:
    print(solution(325489))


if __name__ == "__main__":
    main()

File: 2017/day03-1.ts
function solution(content: number) {
  if (content < 1) throw new Error(`Invalid input: ${content}`);
  if (content === 1) return 0;
  if (content < 10) return content % 2 === 0 ? 1 : 2;
  const sqrt = Math.ceil(Math.sqrt(content));
  const side = sqrt % 2 == 1 ? sqrt : sqrt + 1;
  const half = Math.floor(side / 2);
  const diff = Math.min(
    ...[
      Math.pow(side - 2, 2) + half,
      Math.pow(side - 2, 2) + half + side - 1,
      Math.pow(side - 2, 2) + half + (side - 1) * 2,
      Math.pow(side - 2, 2) + half + (side - 1) * 3,
    ].map((x) => Math.abs(x - content))
  );
  return half + diff;
}

console.log(solution(325489));
